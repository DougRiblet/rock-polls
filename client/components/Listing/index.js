import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { grabAllPolls } from '../../actions/actions';
import listingPresent from './listing-presentational';

const mapStateToProps = state => ({
  allPolls: state.allpolls,
});

const mapDispatchToProps = dispatch => bindActionCreators({ grabAllPolls }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(listingPresent);
