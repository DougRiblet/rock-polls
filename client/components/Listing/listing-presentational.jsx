// @flow

import React from 'react';
import { Link } from 'react-router-dom';

type PollOption = {
  question: string,
  date: string,
};

type PollOptions = {
  [key: string]: PollOption
};

type Props = {
  grabAllPolls: () => mixed,
  allPolls: PollOptions,
};

export default class Listing extends React.Component<Props> {
  constructor(props: Props) {
    super(props);
    this.showListing = this.showListing.bind(this);
  }

  componentDidMount() {
    this.props.grabAllPolls();
  }

  showListing() {
    const ap = Object.entries(this.props.allPolls);
    return ap
      .sort((a, b) => new Date(b[1].date) - new Date(a[1].date))
      .map(p => (
        <li key={p[0]}>
          <Link to={`/polls/${p[0]}`}>{p[1].question}</Link>
          <span className='polldate'> {p[1].date.slice(0, 10)}</span>
        </li>
      ));
  }

  render() {
    const notReady = !this.props.allPolls || !Object.keys(this.props.allPolls).length;
    return (
      <div id='show-all-polls'>
        <div id='listing-head'>
          <h1>Latest Polls</h1>
        </div>
        <div id='listing-list'>
          {
            notReady
            ? <p className='loading'>L O A D I N G</p>
            : <ul>{ this.showListing() }</ul>
          }
        </div>
      </div>
    );
  }
}
