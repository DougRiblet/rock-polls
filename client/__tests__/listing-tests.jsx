import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Listing from '../components/Listing/listing-presentational';

jest.mock('react-router-dom', () => ({ Link: 'Link' }))

describe('Listing', () => {
  test('Snapshot', () => {
    const tree = renderer.create(<Listing
      grabAllPolls={() => jest.fn()}
      allPolls={{}}
    />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Listing
      grabAllPolls={() => jest.fn()}
      allPolls={{}}
    />, div);
  });
});
