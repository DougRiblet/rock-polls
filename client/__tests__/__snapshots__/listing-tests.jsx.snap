// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Listing Snapshot 1`] = `
<div
  id="show-all-polls"
>
  <div
    id="listing-head"
  >
    <h1>
      Latest Polls
    </h1>
  </div>
  <div
    id="listing-list"
  >
    <p
      className="loading"
    >
      L O A D I N G
    </p>
  </div>
</div>
`;
